lib/NavigatorWatcher.js

```diff
@@ -15,66 +15,85 @@
+    waitUntil = waitUntil.map(value => {
+      const protocolEvent = puppeteerToProtocolLifecycle[value];
+      console.assert(protocolEvent, 'Unknown value for options.waitUntil: ' + value);
+      return protocolEvent;
+    });
+
+    this._eventListeners = [];
+    this._frameManager = frameManager;
+    this._frame = frame;
+    this._initialLoaderId = frame._loaderId;
+    this._timeout = typeof options.timeout === 'number' ? options.timeout : 30000;
+    this._waitUntil = waitUntil;
   }
 
   /**
    * @return {!Promise<?Error>}
    */
-  async waitForNavigation() {
-    this._eventListeners = [];
+  _timeoutPromise() {
+    if (!this._timeout)
+      return new Promise(() => {});
+    const errorMessage = 'Navigation Timeout Exceeded: ' + this._timeout + 'ms exceeded';
+    return new Promise(fulfill => this._maximumTimer = setTimeout(fulfill, this._timeout))
+      .then(() => new Error(errorMessage))
+  }
 
-    const navigationPromises = [];
-    if (this._timeout) {
-      const watchdog = new Promise(fulfill => this._maximumTimer = setTimeout(fulfill, this._timeout))
-          .then(() => 'Navigation Timeout Exceeded: ' + this._timeout + 'ms exceeded');
-      navigationPromises.push(watchdog);
-    }
+  /**
+   * @return {!Promise<?Error>}
+   */
+  async waitForNavigation() {
+    this._eventListeners = [
+      helper.addEventListener(this._frameManager, FrameManager.Events.LifecycleEvent, this._checkLifecycleComplete.bind(this))
+    ];
 
-    this._eventListeners.push(helper.addEventListener(this._client, 'Page.lifecycleEvent', this._onLifecycleEvent.bind(this)));
-    const pendingEventsFired = new Promise(fulfill => this._pendingEventsCallback = fulfill);
-    navigationPromises.push(pendingEventsFired);
+    const lifecycleCompletePromise = new Promise(fulfill => this._lifecycleCompleteCallback = fulfill);
 
-    const error = await Promise.race(navigationPromises);
+    this._checkLifecycleComplete();
+    const error = await Promise.race([
+      this._timeoutPromise(),
+      lifecycleCompletePromise
+    ]);
     this._cleanup();
-    return error ? new Error(error) : null;
+    return error;
   }
 
-  /**
-   * @param {!{frameId: string, name: string}} event
-   */
-  _onLifecycleEvent(event) {
-    if (event.frameId !== this._frameId)
+  _checkLifecycleComplete() {
+    // We expect navigation to commit.
+    if (this._frame._loaderId === this._initialLoaderId)
       return;
-    const pptrName = protocolLifecycleToPuppeteer[event.name];
-    if (!pptrName)
+    if (!checkLifecycle(this._frame, this._waitUntil))
       return;
-    this._pendingEvents.delete(pptrName);
-    if (this._pendingEvents.size === 0)
-      this._pendingEventsCallback();
+    this._lifecycleCompleteCallback();
+
+    /**
+     * @param {!Frame} frame
+     * @param {!Array<string>} waitUntil
+     */
+    function checkLifecycle(frame, waitUntil) {
+      let result = true;
+      for (let i = 0; result && i < waitUntil.length; ++i)
```

Comment:
for (var event of waitUntil) {
if (!frame._lifecycleEvents.has(event)) return false;
}

Reply:
This is nice, done!